Incorrect bridge gen for method that has value class with underlying Object type as result type

Incorrect bridge generation for method that has value class with the underlying Object type as result type.
The compilation of this code with -Ycheck:all:

``` scala
class A(val x: Object) extends AnyVal
abstract class VCACompanion[T] {
  def yyyy(x: Object): T
}
class XXX extends VCACompanion[A] {
  override def yyyy(x: Object): A = new A("fff")
}
```

results in:

```
 method yyyy is already defined as method yyyy: (x: Object)ErasedValueType(A, Object)
 (the definitions have matching type signatures)
```

The problem is that there is a bridge generation during Erasure for `def yyyy(x: Object): A = ...`. 
This bridge has the same signature as original `def yyyy(x: Object): A = ...`:

```
override def yyyy(x: Object): Object = "fff"
def yyyy(x: Object): Object = new A(this.yyyy(x))
```

Currently scalac rejects this example (2.11 and 2.12) with:
```
test.scala:8: error: bridge generated for member method produce: ()X in class XProducer
which overrides method produce: ()T in trait Producer
clashes with definition of the member itself;
both have erased type ()Object
  def produce = new X(null)
      ^
one error found
```

OK, I think that is acceptable.

On Fri, Mar 18, 2016 at 5:45 PM, Vladimir Nikolaev <notifications@github.com
> wrote:

> Currently scalac rejects this example (2.11 and 2.12) with:
>
> test.scala:8: error: bridge generated for member method produce: ()X in class XProducer
> which overrides method produce: ()T in trait Producer
> clashes with definition of the member itself;
> both have erased type ()Object
>   def produce = new X(null)
>       ^
> one error found
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/issues/1169#issuecomment-198447068>
>



-- 
Martin Odersky
EPFL

--- a/compiler/src/dotty/tools/dotc/transform/Bridges.scala
+++ b/compiler/src/dotty/tools/dotc/transform/Bridges.scala
@@ -56,7 +56,13 @@ class Bridges(root: ClassSymbol)(implicit ctx: Context) {
       bridgesScope.lookupAll(member.name).exists(bridge =>
         bridgeTarget(bridge) == member && bridge.signature == other.signature)
     def info(sym: Symbol)(implicit ctx: Context) = sym.info
-    def desc(sym: Symbol)= i"$sym${info(sym)(preErasureCtx)} in ${sym.owner}"
+    def desc(sym: Symbol)= {
+      val infoStr = info(sym)(preErasureCtx) match {
+        case ExprType(info) => i": $info"
+        case info => info.show
+      }
+      i"$sym$infoStr in ${sym.owner}"
+    }
     if (member.signature == other.signature) {
       if (!member.info.matches(other.info))
         ctx.error(em"""bridge generated for member ${desc(member)}

--- a/tests/neg/i1169.scala
+++ b/tests/neg/i1169.scala
@@ -1,0 +1,9 @@
+class X(val underlying: Object) extends AnyVal
+
+trait Producer[T] {
+  def produce: T
+}
+
+class XProducer extends Producer[X] {
+  def produce = new X(null) // error
+}

